lldb/source/Host/windows/MainLoopWindows.cpp

```diff
@@ -44,36 +37,61 @@ MainLoopWindows::~MainLoopWindows() {
 }
 
 llvm::Expected<size_t> MainLoopWindows::Poll() {
-  std::vector<WSAEVENT> events;
+  std::vector<HANDLE> events;
   events.reserve(m_read_fds.size() + 1);
-  for (auto &[fd, info] : m_read_fds) {
-    int result = WSAEventSelect(fd, info.event, FD_READ | FD_ACCEPT | FD_CLOSE);
-    assert(result == 0);
-    UNUSED_IF_ASSERT_DISABLED(result);
-
-    events.push_back(info.event);
+  for (auto &[fd, fd_info] : m_read_fds) {
+    // short circuit waiting if a handle is already ready.
+    if (fd_info.object_sp->HasReadableData())
+      return events.size();
+    events.push_back(fd);
   }
   events.push_back(m_interrupt_event);
 
-  DWORD result =
-      WSAWaitForMultipleEvents(events.size(), events.data(), FALSE,
-                               ToTimeout(GetNextWakeupTime()), FALSE);
+  while (true) {
+    DWORD timeout = INFINITY;
+    std::optional<lldb_private::MainLoopBase::TimePoint> deadline =
+        GetNextWakeupTime();
+    if (deadline) {
+      // Check how much time is remaining, we may have woken up early for an
+      // unrelated reason on a file descriptor (e.g. a stat was triggered).
+      std::chrono::milliseconds remaining =
+          std::chrono::duration_cast<std::chrono::milliseconds>(
+              deadline.value() - std::chrono::steady_clock::now());
+      if (remaining.count() <= 0)
+        return events.size() - 1;
+      timeout = remaining.count();
+    }
```

Comment:
Is this any different from
```
DWARF timeout = ToTimeout(GetNextWakeupTime());
if (timeout == 0)
  return events.size() - 1;
```
(using the original definition of the ToTimeout function) ?

Reply:
Reverted this and some of the other changes to MainLoopWindows::Poll, its more like the previous version but we now have additional helpers for other types.